### PR TITLE
Added script to build and upload docker image for matrix test.  Added…

### PR DESCRIPTION
**Pull request by yongni**
… build_interop.sh for go 1.7 and 1.8 from template.

This PR is to collect feedback before I move forward.   The code will need more polish.  

A sample output showing the build and upload of docker images for go.

$ tools/gcp/utils/create_matrix_images.py
START: Building interop docker images.
Jobs to run: 
build_docker_go1.7: TTY_FLAG=-t INTEROP_IMAGE=gcr.io/grpc-testing/grpc_interop_go1.7:latest BASE_NAME=grpc_interop_go1.7 tools/run_tests/dockerize/build_interop_image.sh
build_docker_go1.8: TTY_FLAG=-t INTEROP_IMAGE=gcr.io/grpc-testing/grpc_interop_go1.8:latest BASE_NAME=grpc_interop_go1.8 tools/run_tests/dockerize/build_interop_image.sh

PASSED: build_docker_go1.8 [time=25.5sec; retries=0:0]
PASSED: build_docker_go1.7 [time=26.0sec; retries=0:0]
SUCCESS: All docker images built successfully.
START: Uploading gcr.io/grpc-testing/grpc_interop_go1.7:latest
The push refers to a repository [gcr.io/grpc-testing/grpc_interop_go1.7]
e9b5c39ecdb4: Pushed 
d7e118a496d3: Layer already exists 
4c0afac86c8a: Layer already exists 
c5d3a23082db: Layer already exists 
0846dad56692: Layer already exists 
ecf8ee1b1850: Layer already exists 
4a3a49900086: Layer already exists 
265727c53321: Layer already exists 
e09ae59adef3: Layer already exists 
3cb4a3b0d1a4: Layer already exists 
d359ab38b013: Layer already exists 
682e7cee9d37: Layer already exists 
295d6a056bfd: Layer already exists 
latest: digest: sha256:d102f7b8aec997b86a664e64b7eaedc916a8cde25b29db1b330b4aa5b58dbd42 size: 3057
START: Uploading gcr.io/grpc-testing/grpc_interop_go1.8:latest
The push refers to a repository [gcr.io/grpc-testing/grpc_interop_go1.8]
13d9d7012d6a: Pushed 
04d375f7a66a: Layer already exists 
bb9b4ab56a67: Layer already exists 
735cb668e99c: Layer already exists 
c7de9736dfd0: Layer already exists 
437d17e16e98: Layer already exists 
791933004643: Layer already exists 
df292491f3b9: Layer already exists 
ee40cbc5082f: Layer already exists 
3cb4a3b0d1a4: Layer already exists 
d359ab38b013: Layer already exists 
682e7cee9d37: Layer already exists 
295d6a056bfd: Layer already exists 
latest: digest: sha256:61eb17f4e99381103ea03a47207f2ea29bc0a6e0b7daa1455d031a003415e89a size: 3057
yongni@canaan:~/github/grpc$

### CODE REVIEW

File: templates/tools/dockerfile/go_build_interop.sh [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
```
Comment (jtattermusch) [inline templates/tools/dockerfile/go_build_interop.sh]: nit: please add .include suffix (all other included files do thathttps://github.com/grpc/grpc/tree/master/templates/tools/dockerfile)

File: tools/gcp/utils/create_matrix_images.py [hunk 2]
```diff
@@ -28,25 +28,39 @@
 # (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-"""Upload docker images to Google Container Registry."""
+"""Build and upload docker images to Google Container Registry per matrix."""
```
Comment (jtattermusch) [inline tools/gcp/utils/create_matrix_images.py]: tools/gcp/utils  (read as "utilities for Google computing platform") is a wrong place for these scripts IMHO.
Do you want to create  e.g. tools/interop_matrix  (or a different name, but just "matrix" is not specific enough) where all these scripts will live?  You can add a tools/interop_matrix/README.md that explains what you are trying to achieve.

File: tools/gcp/utils/create_matrix_images.py [hunk 3]
```diff
@@ -94,26 +110,86 @@ def upload_to_gcr(image):
...
 
+  jobset.message('START', 'Repackaging %s' % image, do_newline=True)
   build_cmd = ['docker', 'build', '--rm', '--tag', new_tag, temp_dir]
   subprocess.check_output(build_cmd)
+  return new_tag
 
-  if not args.keep:
-    atexit.register(lambda: subprocess.call(['docker', 'rmi', new_tag]))
+def upload_to_gcr(image):
+  """Tags and Pushes a docker image in Google Containger Registry.
```
Comment (jtattermusch) [inline tools/gcp/utils/create_matrix_images.py]: typo: Containger

File: templates/tools/dockerfile/go_build_interop.sh (hunk 1 above)
Comment (yongni) [inline templates/tools/dockerfile/go_build_interop.sh]: Done.

File: tools/gcp/utils/create_matrix_images.py (hunk 2 above)
Comment (yongni) [inline tools/gcp/utils/create_matrix_images.py]: Done.

File: tools/gcp/utils/create_matrix_images.py (hunk 3 above)
Comment (yongni) [inline tools/gcp/utils/create_matrix_images.py]: Done.

**Comment by yongni** [comment]
See the update that included the go__latest test case and how it is generated.

See sample output from create_testcases.sh:

yongni@canaan:~/github/grpc/tools/interop_matrix$ LANG=go ./create_testcases.sh 
Create 'go' test cases for gRPC release 'latest'
Seen --use_docker flag, will run interop tests under docker.

IMPORTANT: The changes you are testing need to be locally committed
because only the committed changes in the current branch will be
copied to the docker environment.
START: Building interop docker images.
PASSED: build_docker_go [time=24.0sec; retries=0:0]
SUCCESS: All docker images built successfully.
All tests will skipped --manual_run option is active.
SKIPPED: cloud_to_prod:default:go:large_unary
SKIPPED: cloud_to_prod:default:go:empty_unary
SKIPPED: cloud_to_prod:default:go:ping_pong
SKIPPED: cloud_to_prod:default:go:empty_stream
SKIPPED: cloud_to_prod:default:go:client_streaming
SKIPPED: cloud_to_prod:default:go:server_streaming
SKIPPED: cloud_to_prod:default:go:cancel_after_begin
SKIPPED: cloud_to_prod:default:go:cancel_after_first_response
SKIPPED: cloud_to_prod:default:go:timeout_on_sleeping_server
SUCCESS: All tests passed
Command log written to file interop_client_cmds.sh
Preserving docker image: grpc_interop_go:d84d1fa8-a318-4987-a1ac-65ee6c2d6c1e
Running the test cases:
Testing grpc_interop_go:d84d1fa8-a318-4987-a1ac-65ee6c2d6c1e
2017/05/15 17:50:58 LargeUnaryCall done
2017/05/15 17:51:02 EmptyUnaryCall done
2017/05/15 17:51:06 Pingpong done
2017/05/15 17:51:09 Emptystream done
2017/05/15 17:51:12 ClientStreaming done
2017/05/15 17:51:16 ServerStreaming done
2017/05/15 17:51:18 CancelAfterBegin done
2017/05/15 17:51:21 CancelAfterFirstResponse done
2017/05/15 17:51:24 TimeoutOnSleepingServer done
Updating: /usr/local/google/home/yongni/github/grpc/tools/interop_matrix/testcases/go__latest
2c2
< echo "Testing ${docker_image:=grpc_interop_go:bdc16235-cba6-4d9a-abb9-0b2713d147a7}"
---
> echo "Testing ${docker_image:=grpc_interop_go:d84d1fa8-a318-4987-a1ac-65ee6c2d6c1e}"
Test cases created: /usr/local/google/home/yongni/github/grpc/tools/interop_matrix/testcases/go__latest
testcase: /usr/local/google/home/yongni/github/grpc/tools/interop_matrix/testcases/go__latest
Clean up docker_image grpc_interop_go:d84d1fa8-a318-4987-a1ac-65ee6c2d6c1e
Untagged: grpc_interop_go:d84d1fa8-a318-4987-a1ac-65ee6c2d6c1e
Deleted: sha256:d567fb5ab0db168c86b164cc8c5cd2744102cafe3ebba3f0de650205756f5605
Deleted: sha256:f7f8329bd847be061520bf7e5ec033e80e9c5e02dfdf1bf3751bc75a9269386b

**Comment by yongni** [comment]
@jtattermusch @makdharma @adelez @jayantkolhe:

With the latest commit to my matrix branch, we are mostly feature complete in term of building docker image and creating test cases.  As of now, with single command invocation, the framework could build docker image for all languages for all gRPC releases (referencing git tag) across all defined runtimes and upload these images to GCR.

We still need to resolve openjdk issue tracked in #11071 before we cover java runtimes.  I need to also polish the code more and working on the test automation part.  However, I will like to merge to master before continue on the latter part.  Hence, your feedback of this PR is appreciated.

The sample output below shows how the script builds and uploads
   [go1.7, go1.8] X [v1.0.1-GA, v1.3.0], total 4 images.
If we want to add go1.6 or other releases, it would be 1 line change in the client_matrix.py file.  See tools/interop_matrix/README.md for details.

[create_matrix_images.output.txt](https://github.com/grpc/grpc/files/1009873/create_matrix_images.output.txt)

**Comment by yongni** [comment]
Known failures:
Bazel Full Build & Tests - Ignored?
Portability Tests Linux - #10881
gRPC_interop_pull_requests - #11242

**Comment by stanley-cheung** [comment]
Why are we moving the `third_party/protobuf` commit has from `v3.3.0` back to `v3.2.0`, in this PR?

**Comment by stanley-cheung** [comment]
See #10995 cc @kpayson64

**Comment by matt-kwong** [comment]
It looks like a mistake that was made when this branch was merged before rebasing to master.

**Pull request merged**